File: core/scheduler.py
"""
TinyOLED Desktop — Görev Zamanlayıcı
Periyodik görevleri (saat güncelleme, WiFi kontrol vs.)
ana döngüyü bloke etmeden arka planda çalıştırır.
"""

import threading
import time
from typing import Callable, Dict, Tuple


class Task:
    def __init__(self, name: str, fn: Callable, interval: float):
        self.name     = name
        self.fn       = fn
        self.interval = interval
        self.last_run = 0.0
        self.enabled  = True

    def is_due(self, now: float) -> bool:
        return self.enabled and (now - self.last_run) >= self.interval

    def run(self):
        try:
            self.fn()
        except Exception as e:
            print(f"[SCHED] '{self.name}' hatası: {e}")
        self.last_run = time.monotonic()


class Scheduler:
    """
    Hafif kooperatif zamanlayıcı.
    Ana döngüde her frame `tick()` çağrılır;
    vadesi gelmiş görevler sırayla çalıştırılır.
    """

    def __init__(self):
        self._tasks: Dict[str, Task] = {}

    def add(self, name: str, fn: Callable, interval: float):
        """
        Görev ekle.
        interval: saniye cinsinden çalışma aralığı.
        """
        self._tasks[name] = Task(name, fn, interval)

    def remove(self, name: str):
        self._tasks.pop(name, None)

    def enable(self, name: str, state: bool = True):
        if name in self._tasks:
            self._tasks[name].enabled = state

    def tick(self):
        """Her frame çağrılır; vadesi gelen görevleri çalıştırır."""
        now = time.monotonic()
        for task in list(self._tasks.values()):
            if task.is_due(now):
                task.run()

    def run_once(self, name: str):
        """Bir görevi hemen (vadesi gelmeden) tetikle."""
        task = self._tasks.get(name)
        if task:
            task.run()

File: web-simulator/js/scheduler.js
/**
 * TinyOLED Desktop — Cooperative Task Scheduler (JavaScript Port)
 * Runs periodic tasks within the requestAnimationFrame loop.
 */

export class Scheduler {
  constructor() {
    this._tasks = new Map();
  }

  add(name, fn, interval) {
    this._tasks.set(name, {
      name,
      fn,
      interval,          // seconds
      lastRun: 0,
      enabled: true,
    });
  }

  remove(name) {
    this._tasks.delete(name);
  }

  enable(name, state = true) {
    const t = this._tasks.get(name);
    if (t) t.enabled = state;
  }

  tick() {
    const now = performance.now() / 1000;
    for (const task of this._tasks.values()) {
      if (task.enabled && (now - task.lastRun) >= task.interval) {
        try { task.fn(); } catch (e) {
          console.error(`[SCHED] '${task.name}' error:`, e);
        }
        task.lastRun = now;
      }
    }
  }
}
